[SROA] Canonicalize homogeneous structs to fixed vectors (opt-in, after memcpyopt)

yxsamliu opened the pull request:
SROA sometimes keeps temporary allocas around for homogeneous structs like
`{ i32, i32, i32, i32 }` because the partition has only memcpy/memset traffic
and no scalar typed users to drive vector promotion. On targets like AMDGPU
these allocas turn into scratch memory and hurt performance. This PR adds a
helper `tryCanonicalizeStructToVector` that converts such a partition to a
fixed vector type when every non-debug, non-lifetime user is a memory
intrinsic, so the alloca can promote through normal vector load/store paths.
The element-shape rule accepts any homogeneous element count, any integer
width, any FP type, and integral pointer types, as long as the struct is
tightly packed.

Canonicalization is gated behind a new per-pass option
`canonicalize-struct-to-vector` on `SROAOptions`, off by default. Only the
late SROA passes in `addVectorPasses` (new PM, non-LTO and FullLTO) and the
two legacy-PM SROAs in NVPTX enable it, so it always runs after
`MemCpyOptPass`. Running it earlier can hide memcpy chains that memcpyopt
would otherwise collapse, and can also emit wide stores whose suffix lanes
are undef when only part of a struct was initialized. Both hazards are
covered by new tests `struct-to-vector-before-memcpyopt.ll` and
`struct-to-vector-fp-store-only-tail.ll`. The opt-in design and the
"after memcpyopt" placement come from @YonahGoldberg's refactor, which
removed every regression reported in earlier benchmark runs (see
dtcxzyw/llvm-opt-benchmark-nightly#306). AMDGPU inherits the new-PM opt-ins
automatically; other targets keep upstream-main SROA behavior unless they
opt in.

Co-authored-by: Yonah Goldberg <ygoldberg@nvidia.com>

yxsamliu commented:
ping

yxsamliu commented:
Thanks for offering to review! I've rebased onto latest main and fixed the semantic conflict (`SliceTy` → `PartitionTy` rename). Also regenerated the CHECK lines using `update_test_checks.py`. Ready for your review.

YonahGoldberg reviewed (commented):
Overall looks really good! I do think it can be generalized though, but I'm probably fine to merge it after you respond to a couple of my nits and we can generalize it later (unless you want to work on that then go ahead). I'm pretty new to upstream contributions, so let's get some second opinions from the more experienced people you asked to review as well.

yxsamliu commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5329,6 +5396,15 @@ AllocaInst *SROA::rewritePartition(AllocaInst &AI, AllocaSlices &AS,
   auto [PartitionTy, IsIntegerWideningViable, VecTy] =
       selectPartitionType(P, DL, AI, *C);
 
+  // Canonicalize homogeneous, tightly-packed structs to vector types.
```
Done. Moved the `tryCanonicalizeStructToVector()` call inside `selectPartitionType()`, right before the fallback return as you suggested. Also added `MaxStructToVectorSize` as a parameter to pass it through.

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
The TTI issue will also go away if we remove the pass option. I'm still waiting to see llvm-opt-benchmark results with the struct size heuristic removed.

yxsamliu commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -546,6 +546,13 @@ class TargetTransformInfo {
+  /// canonicalize to a vector type. This enables better optimization of
+  /// tightly-packed structs on targets where scratch memory is expensive.
+  ///
+  /// \returns 0 to disable the transformation, or the maximum struct size.
```
sorry for the delay. I ran the benchmark locally. I have updated the PR to remove the parameter and posted this PR to https://github.com/dtcxzyw/llvm-opt-benchmark/issues/1312

yxsamliu replied:
The benchmark results are in: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/3475. 1560 files changed with 10K fewer lines overall. `sroa.NumVectorized` went up by 0.95%. The main regression is `memcpyopt.NumCpyToSet` at -11.29%, which is expected — when SROA vectorizes a struct, the memcpy goes away so memcpyopt has fewer things to optimize. Everything else is under 1%. Looks clean to me. I've removed the size limit option and the TTI dependency as discussed.

nikic reviewed (commented):
It looks like this is introducing a new heuristic that is based on the alloca type. @vtjnash is currently actively working on removing dependencies on the alloca type, so I think we should avoid this.

Can this information be derived based on access patterns instead? In the motivating case, it looks like there are i64 stores on which this can be based?

yxsamliu commented:
> > It looks like this is introducing a new heuristic that is based on the alloca type. @vtjnash is currently actively working on removing dependencies on the alloca type, so I think we should avoid this.
> > Can this information be derived based on access patterns instead? In the motivating case, it looks like there are i64 stores on which this can be based?
> 
> I think it might be feasible. Let me try that.

removed the TypePartitionTy == AI.getAllocatedType() check

vtjnash commented:
> removed the TypePartitionTy == AI.getAllocatedType() check

The request nikic was making, on my behalf, was to pull this entirely outside of the `AI.getAllocatedType()` branch of the function, since I'm moving towards being able to delete that branch, as he mentioned. That said, I don't think that change is actually necessary for this PR. As long as you're confident in your tests here representing a cross-section of what this optimization should be able to do, then I think it will provide good coverage to ensure any future changes preserve the intent here.

vtjnash commented on llvm/test/Transforms/SROA/struct-to-vector-subpartition.ll:
```diff
@@ -0,0 +1,45 @@
+; RUN: opt -passes=sroa -S %s | FileCheck %s
```
This should have an auto-generated check lines if possible, per
https://llvm.org/docs/TestingGuide.html#generating-assertions-in-regression-tests

Or a comment "NOTE: Do not autogenerate" with an explanation why it won't work

yxsamliu replied:
Added a NOTE comment explaining why. The test uses a broad CHECK-NOT for `<2 x i64>` to guard against type propagation in the fallback path, which doesn't work well with autogenerated checks.

YonahGoldberg commented:
I'll look at the changes tomorrow, at first glance it looks good though. Could you confirm this is the latest benchmark run? https://github.com/dtcxzyw/llvm-opt-benchmark/pull/3520. Looks like really good improvements!

yxsamliu commented:
> > > > I'll look at the changes tomorrow, at first glance it looks good though. Could you confirm this is the latest benchmark run? [dtcxzyw/llvm-opt-benchmark#3520](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/3520). Looks like really good improvements!
> > > 
> > > 
> > > That one is pretty close but I made some minor changes after that but forgot to trigger a new run. I just triggered a new run and let's see how it goes: [dtcxzyw/llvm-opt-benchmark#3547](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/3547)
> > 
> > 
> > Hmm unfortunately looks like there's one regression that doesn't look good. We're vectorizing when the alloca gets promoted as a scalar anyway, but then this requires a bunch of inserts and extracts.
> > Sorry to be such a stickler, again I haven't reviewed much so IDK exactly what LLVM policy is, but I assume there should be a pretty high bar for transforms like SROA because it's such an important pass.
> 
> I am OK. I think we should try fixing any regressions caused by changes in llvm passes since that could be a tip of an iceberg. I will look into this. Thanks.

got a fix for that regression and triggered a new run https://github.com/dtcxzyw/llvm-opt-benchmark/pull/3556

the regression happens when all uses of the partition are by components therefore no benefit of promoting to vector type. just added detection for that case and excluded it

yxsamliu committed:
[SROA] Canonicalize homogeneous structs to vectors with guarded fallback

When SROA's getTypePartition yields a homogeneous struct (no padding, no
pointers, no i1 fields), canonicalize it to a fixed vector (e.g.
{ i64, i64 } -> <2 x i64>) so allocas can promote through vector load/store
patterns such as std::function swap with three memcpys.

Introduce shouldCanonicalizeHomogeneousStructToVector for the struct-to-vector
fallback after normal promotion paths fail: require a non-splittable
whole-partition use, reject non-splittable sub-element loads, and for i64
homogeneous partitions recover splittable MemIntrinsic transfers only for
interior subaggregates or full-record integer aggregates of at least 32 bytes
(avoiding the broad 16-byte memcpy bucket that reopened llvm-opt-benchmark
regressions).

Add LLVM_DEBUG tracing for partition-type decisions, extend SROA lit coverage,
update DebugInfo SROA FileChecks for struct-to-vector codegen, and regenerate
NVPTX lower-byval-args.ll expectations (@memcpy_to_param) using llc from
llvm-dev Docker.

yxsamliu commented:
@YonahGoldberg 

I updated the heuristic to try to keep the useful performance wins from vectorizing alloca partitions, while avoiding the cases that looked like they were causing regressions, and I triggered a new llvm-opt-benchmark run for that latest change here: 

https://github.com/dtcxzyw/llvm-opt-benchmark-nightly/pull/152

The IR report there still shows a few apparent regressions, so I followed up locally by checking the corresponding assembly for those changed files, and I could not find a clear asm-level performance regression signal. Most of the changed cases looked neutral to better in asm, and the one remaining small positive asm delta in `cpython/compile` still looked more like control-flow / register-allocation noise than a real slowdown. So my current reading is that the updated heuristic is keeping the vectorization benefit we want, and that the apparent regressions in the IR diff do not look like real performance regressions from the assembly side.

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+  if (StructSize != NumElts * EltSize)
+    return nullptr;
+
+  // Verify each element is at the expected offset (no padding).
```
I think this check is already implied by the above check that `size == NumElts * EltSize` and by the fact that all elements are the same size.

yxsamliu replied:
will remove redundant checks

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+    if (STy->getElementType(I) != EltTy)
+      return nullptr;
+
+  // Element type must be valid for vectors.
```
I think this check isn't necessary because below already checks that this is an integer type >= 8 bits or floating point type, which is stricter than than `VectorType::isValidElementType`

yxsamliu replied:
will remove

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+
+  const StructLayout *SL = DL.getStructLayout(STy);
+  uint64_t StructSize = SL->getSizeInBytes();
+  if (StructSize == 0)
```
I don't think the struct size can be zero because we are already checking that the struct has 2 or 4 elements and those elements are floating point or a >=8 bit sized integer and we check `size == NumElts * EltSize` below.

yxsamliu replied:
removed the separate zero-size guard and now just rely on the existing element checks together with the packed-size check below

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+  if (!EltTS.isFixed())
+    return nullptr;
+  uint64_t EltSize = EltTS.getFixedValue();
+  if (EltSize < 1)
```
EltSize can't be < 1 because we already check it's a floating point type or an integer with bit width >= 8.

yxsamliu replied:
will remove

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+  Type *EltTy = STy->getElementType(0);
+  for (unsigned I = 1; I < NumElts; ++I)
+    if (STy->getElementType(I) != EltTy)
+      return nullptr;
```
```
if !(llvm::all_equal(STy->elements())
```

yxsamliu replied:
will do

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5130,6 +5130,162 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+/// \param STy The struct type to try to canonicalize.
+/// \param DL The DataLayout for size/alignment queries.
+/// \returns The equivalent vector type, or nullptr if not applicable.
+static FixedVectorType *tryCanonicalizeStructToVector(StructType *STy,
```
So should it be:

```
static FixedVectorType *tryCanonicalizeStructToVector(StructType *STy,
                                                      const DataLayout &DL) {
  unsigned NumElts = STy->getNumElements();
  if (NumElts != 2 && NumElts != 4)
    return nullptr;

  Type *EltTy = STy->getElementType(0);
  if (!all_equal(STy->elements()))
    return nullptr;
  if (auto *IT = dyn_cast<IntegerType>(EltTy)) {
    if (IT->getBitWidth() < 8)
      return nullptr;
  } else if (!EltTy->isFloatingPointTy()) {
    return nullptr;
  }

  TypeSize EltTS = DL.getTypeAllocSize(EltTy);
  if (!EltTS.isFixed())
    return nullptr;
  uint64_t EltSize = EltTS.getFixedValue();
  if (DL.getStructLayout(STy)->getSizeInBytes() != NumElts * EltSize)
    return nullptr;
  return FixedVectorType::get(EltTy, NumElts);
}
```

yxsamliu replied:
Yes, I moved it in that direction. I simplified the structural part of tryCanonicalizeStructToVector along those lines, and I also folded the conservative fallback decision into that same helper so the logic is not split across two separate functions anymore

YonahGoldberg commented on llvm/test/Transforms/SROA/struct-to-vector-mapok-extra-alloca.ll:
```diff
@@ -0,0 +1,35 @@
+; RUN: opt -passes='default<O3>' -S %s | FileCheck %s
```
Have you run llvm-reduce on this? Also, I think we usually wouldn't like to run the whole pass pipeline on lit tests, we'd probably want to extract what the IR look like right before SROA. Also, the names here are very verbose. Also, is this target dependent? I think we usually remove the datalayout and triple if it's. not necessary for the check.

yxsamliu replied:
removed the broad default<O3> test and kept the focused -passes=sroa tests that directly cover the current fallback boundaries

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5144,6 +5300,26 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+        dbgs() << " chosen=" << *ChosenTy;
+      if (ChosenVecTy)
+        dbgs() << " vec=" << *ChosenVecTy;
+      dbgs() << " intwiden=" << ChosenIntWidening << "\n";
```
int-widen

yxsamliu replied:
fixed

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5144,6 +5300,26 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
 static std::tuple<Type *, bool, VectorType *>
 selectPartitionType(Partition &P, const DataLayout &DL, AllocaInst &AI,
                     LLVMContext &C) {
+  auto LogChoice = [&](StringRef Path, Type *ChosenTy, VectorType *ChosenVecTy,
```
This is very useful for debugging, thanks!

YonahGoldberg replied:
nit: can we rename to LogSelection and SelectedTy

yxsamliu replied:
Sure, I renamed that helper to LogSelection and used the corresponding Selected* names there to make it a little clearer.

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5192,32 +5373,83 @@ selectPartitionType(Partition &P, const DataLayout &DL, AllocaInst &AI,
+        //
+        // So this path is restricted to 64-bit integer lanes, and only for
+        // interior partitions or original full records of size >= 32 bytes.
+        bool AllowStructFallback = shouldCanonicalizeHomogeneousStructToVector(
```
I think it's confusing to have this as a separate function and then you have a notion of "fallback recovery", which is verbose. I'd just have one function `tryCanonicalizeStructToVector` that does all the decision making and returns the vector type on success.

yxsamliu replied:
folded that logic into tryCanonicalizeStructToVector

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5192,32 +5373,83 @@ selectPartitionType(Partition &P, const DataLayout &DL, AllocaInst &AI,
+        // interior partitions or original full records of size >= 32 bytes.
+        bool AllowStructFallback = shouldCanonicalizeHomogeneousStructToVector(
+            P, DL, AI, VTy->getElementType()->isIntegerTy(64));
+        LLVM_DEBUG({
```
I think most of this debug info is redundant with the `LogChoice` that gets printed either way.

yxsamliu replied:
removed that extra debug-only block and now just rely on the main selection log

YonahGoldberg reviewed (commented):
My overall feelings are that the heuristic used in `shouldCanonicalizeHomogeneousStructToVector` is overly complex and possibly over-fitted to the llvm-opt-benchmark set. For example, the heuristics on lines 5263 to 5273 seem pretty arbitrary. Why only allow i64 non-whole-alloca memcpys to get transformed? And then why do we disallow canonicalization if there is a sub-element load, but not a sub-element store?

My intuition is the effect you're trying to have here is that you don't want this transformation to result in any `insertelement` or `extractelement` instructions. Is that right? So I thought we could do something much simpler, like say that we can canonicalize if:

1. All the users of the partition are `MemInst`
2. Or all the users are whole loads/stores to the entire partition.

I checked this simpler heuristic against two of your lit tests and found some interesting results:

1. For struct-to-vector-subpartition.ll, this simpler heuristic removes the `alloca { i64, i64 }`. This would get removed later anyway by `memcpyopt`, but I still see this as a win.
2. For struct-to-vector-fp-store-only-tail, this test case is much more interesting. If you're allowed to canonicalize the [0,16] partition to <4 x float> and then you run instcombine, you end up with:

```
define void @store_only_fp_tail(ptr noalias %dst) {
  store <4 x float> <float 0.000000e+00, float undef, float undef, float undef>, ptr %dst, align 1
  %.sroa.3.0.dst.sroa_idx = getelementptr inbounds nuw i8, ptr %dst, i64 16
  store float 0.000000e+00, ptr %.sroa.3.0.dst.sroa_idx, align 1
  ret void
}
```

And then memcpyopt optimizes to:
```
define void @store_only_fp_tail(ptr noalias writeonly captures(none) initializes((0, 20)) %dst) local_unnamed_addr #0 {
  tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 1 dereferenceable(20) %dst, i8 0, i64 20, i1 false)
  ret void
}
```

In comparison, with your heuristic, we disallow the canonicalization and get:
```
define void @store_only_fp_tail(ptr noalias %dst) {
  store float 0.000000e+00, ptr %dst, align 1
  %.sroa.33.0.dst.sroa_idx = getelementptr inbounds nuw i8, ptr %dst, i64 16
  store float 0.000000e+00, ptr %.sroa.33.0.dst.sroa_idx, align 1
  ret void
}
```

So the bug looks like memcpyopt is initializing previously uninitialized memory, so you end up with a wider store, which translates to additional instructions on some backends. I remember finding a similar bug a while back--I wonder if there should be some instcombine transformation that transforms the store <4 x float> to a scalar store? Or should memcpyopt handle this?

Maybe you already saw this and there's a reason you went for a more complicated heuristic. Would be happy to get on a call sometime next week to talk about it if you want because this is getting quite complicated.

yxsamliu committed:
[SROA] Simplify struct-to-vector fallback helper

Fold the fallback decision into tryCanonicalizeStructToVector and drop
redundant structural checks and debug noise so the current conservative
policy is easier to read and review.

yxsamliu commented:
> My overall feelings are that the heuristic used in `shouldCanonicalizeHomogeneousStructToVector` is overly complex and possibly over-fitted to the llvm-opt-benchmark set. For example, the heuristics on lines 5263 to 5273 seem pretty arbitrary. Why only allow i64 non-whole-alloca memcpys to get transformed? And then why do we disallow canonicalization if there is a sub-element load, but not a sub-element store?
> 
> My intuition is the effect you're trying to have here is that you don't want this transformation to result in any `insertelement` or `extractelement` instructions. Is that right? So I thought we could do something much simpler, like say that we can canonicalize if:
> 
> 1. All the users of the partition are `MemInst`
> 2. Or all the users are whole loads/stores to the entire partition.
> 
> I checked this simpler heuristic against two of your lit tests and found some interesting results:
> 
> 1. For struct-to-vector-subpartition.ll, this simpler heuristic removes the `alloca { i64, i64 }`. This would get removed later anyway by `memcpyopt`, but I still see this as a win.
> 2. For struct-to-vector-fp-store-only-tail, this test case is much more interesting. If you're allowed to canonicalize the [0,16] partition to <4 x float> and then you run instcombine, you end up with:
> 
> ```
> define void @store_only_fp_tail(ptr noalias %dst) {
>   store <4 x float> <float 0.000000e+00, float undef, float undef, float undef>, ptr %dst, align 1
>   %.sroa.3.0.dst.sroa_idx = getelementptr inbounds nuw i8, ptr %dst, i64 16
>   store float 0.000000e+00, ptr %.sroa.3.0.dst.sroa_idx, align 1
>   ret void
> }
> ```
> 
> And then memcpyopt optimizes to:
> 
> ```
> define void @store_only_fp_tail(ptr noalias writeonly captures(none) initializes((0, 20)) %dst) local_unnamed_addr #0 {
>   tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 1 dereferenceable(20) %dst, i8 0, i64 20, i1 false)
>   ret void
> }
> ```
> 
> In comparison, with your heuristic, we disallow the canonicalization and get:
> 
> ```
> define void @store_only_fp_tail(ptr noalias %dst) {
>   store float 0.000000e+00, ptr %dst, align 1
>   %.sroa.33.0.dst.sroa_idx = getelementptr inbounds nuw i8, ptr %dst, i64 16
>   store float 0.000000e+00, ptr %.sroa.33.0.dst.sroa_idx, align 1
>   ret void
> }
> ```
> 
> So the bug looks like memcpyopt is initializing previously uninitialized memory, so you end up with a wider store, which translates to additional instructions on some backends. I remember finding a similar bug a while back--I wonder if there should be some instcombine transformation that transforms the store <4 x float> to a scalar store? Or should memcpyopt handle this?
> 
> Maybe you already saw this and there's a reason you went for a more complicated heuristic. Would be happy to get on a call sometime next week to talk about it if you want because this is getting quite complicated.

I cleaned up the current implementation first so the code is easier to review, but I agree the heuristic itself should probably become simpler. I’m pushing this cleanup-only revision first and rerunning llvm-opt-benchmark on it to make sure these local changes are not changing behavior. After that I plan to rework the heuristic in the simpler direction you suggested and see which cases should instead be handled in downstream passes.

yxsamliu committed:
[SROA] Simplify struct-to-vector fallback policy

Keep the fallback rule focused on simple memory-use shapes so downstream passes can handle case-specific codegen fallout.

yxsamliu commented:
@YonahGoldberg I tried the simpler heuristic and looked through the llvm-opt-benchmark diffs it opens up. The only case I found that looks like a real downstream code-shape problem is the assimp/glTFImporter / ImportNode one. There, SROA creates a vector-shaped store that blocks SLP from seeing the full scalar store chain, and SLP ends up with a mixed <4 x float> / <2 x float> / scalar layout. I split that out into a small SLP fix here: https://github.com/llvm/llvm-project/pull/194970.

For the other asm diffs I checked, I do not see enough evidence to call them real regressions right now. A lot of them are XMM spill/reload or vector-copy shape changes. The old code usually had an equivalent scalar/memcpy-style copy sequence. In the simple model, an XMM snapshot across a clobbering call is one 16-byte load, one 16-byte spill store, one 16-byte reload, and one 16-byte final store. The scalar copy shape is two 8-byte loads, two 8-byte temporary stores, two 8-byte reloads, and two 8-byte final stores. Both move 64 bytes total, and the XMM form is not obviously worse by load/store count alone. A small local microbenchmark also showed mixed results depending on the source/destination shape. There are likely more complicated cases than that microbenchmark covers, so I think it is better to treat these as watch items unless we see a concrete cycle-count or real-application regression. So my current plan is to keep the SROA heuristic in the simpler direction you suggested, handle the clear SLP issue separately, and not add more SROA-side profitability rules for the remaining diffs.

yxsamliu committed:
[SROA] Fix struct-to-vector helper formatting

Wrap the LogSelection lambda parameters as requested by clang-format.

YonahGoldberg commented:
Sounds good, let me know when you put the simpler heuristic and please run llvm-opt-benchmark up here so I can take a look at those regressions you're talking about. I don't have experience with SLP so I can't review that change.

What do you think of my change here: https://github.com/llvm/llvm-project/pull/194042#discussion_r3155783298

I think you might run into this issue with the simpler heuristic. Technically the transformation is not legal though...

alexey-bataev commented:
How does it affect targets with the expensive vector-to-scalar/scalar-to-vector data movements?

YonahGoldberg commented:
> How does it affect targets with the expensive vector-to-scalar/scalar-to-vector data movements?

The goal here I believe is to not perform this transformation if it results in any extra `extractelement `or `insertelement`

yxsamliu committed:
[SROA] Fix test updates for simplified fallback

Avoid adding a literal undef CHECK and update the NVPTX byval PTX output after the simplified fallback changes the memcpy_to_param store shape.

yxsamliu commented:
> Sounds good, let me know when you put the simpler heuristic and please run llvm-opt-benchmark up here so I can take a look at those regressions you're talking about. I don't have experience with SLP so I can't review that change.
> 

I pushed the simpler heuristic and reran llvm-opt-benchmark here: https://github.com/dtcxzyw/llvm-opt-benchmark-nightly/pull/261

I generated assembly locally for the baseline and optimized IR for the positive IR deltas in the benchmark, and compared the assembly output. In most cases, I do not see evidence of generated-assembly degradation. Several positive IR deltas actually shrink in assembly, especially the duckdb cases like duckdb/format, duckdb/ub_duckdb_common, and duckdb/ub_duckdb_core_functions_generic.

The main cases that still look worth following up are:

assimp/glTFImporter: this looks like an SLP tiling issue. SROA creates a vector-shaped value, and later SLP ends up with a mixed `<4 x float>` / `<2 x float>` / scalar store shape instead of full `<4 x float>` row stores. I split that out as PR #194970.

node/zstd_compress and duckdb/zstd_compress: these look like X86-specific codegen issues. The vectorized IR uses small `<2 x i32>` values, but the code mostly uses the fields as scalar integers, so X86 gets extra XMM/GPR moves. I think this is better handled in a downstream target-aware pass rather than by making SROA target-specific. This concern is target-specific; AMDGPU and NVPTX do not have the same XMM/GPR register-file boundary.

> What do you think of my change here: [#194042 (comment)](https://github.com/llvm/llvm-project/pull/194042#discussion_r3155783298)
> 

For your PR #194042, I looked through it and I think it is a useful mitigation for the SROA case that exposed the issue. I posted a comment to it.

yxsamliu committed:
[SROA] Update DebugInfo tests for struct-to-vector sub-partition canonicalization

2-element homogeneous struct sub-partitions (e.g. the [x,y] slice of a 3-element
struct, or a struct.two / struct.p field) are now canonicalized to fixed vectors
by tryCanonicalizeStructToVector. Update three DebugInfo tests whose CHECK lines
expected scalar i64/i32 fragments but now see <2 x i64> / <2 x i32> debug values.

yxsamliu committed:
[SROA] Avoid undef in sroasplit-4.ll CHECK pattern

Use {{.*}} instead of literal undef to satisfy the undef deprecator.

YonahGoldberg commented:
@yxsamliu please see https://github.com/yxsamliu/llvm-project/pull/14/changes where I refactored this to eliminate regressions. The trick was to only run this optimization after memcpyopt. I wrote a lot of comments explaining why in the code. There might be some small nitpicks or cleanup after the fact, but I think what I did is directionally right.

See https://github.com/dtcxzyw/llvm-opt-benchmark-nightly/pull/306 for the benchmark results with no regressions.

yxsamliu commented:
> @yxsamliu please see https://github.com/yxsamliu/llvm-project/pull/14/changes where I refactored this to eliminate regressions. The trick was to only run this optimization after memcpyopt. I wrote a lot of comments explaining why in the code. There might be some small nitpicks or cleanup after the fact, but I think what I did is directionally right.
> 
> See [dtcxzyw/llvm-opt-benchmark-nightly#306](https://github.com/dtcxzyw/llvm-opt-benchmark-nightly/pull/306) for the benchmark results with no regressions.

@YonahGoldberg Thanks for the refactor. The design looks right. I ran your branch locally — SROA lit passes, #306 looks clean, and the AMDGPU case that motivated this PR still benefits.

Happy to land it either way. I can cherry-pick your commit onto #165159 (you stay as author), fix the three failing tests, and squash with you as Co-authored-by. Or you land #197087 and we close this one. My priority is getting the fix in. Which do you prefer?

YonahGoldberg commented:
> I can cherry-pick your commit onto https://github.com/llvm/llvm-project/pull/165159 (you stay as author), fix the three failing tests, and squash with you as Co-authored-by.

Sure go ahead and cherry-pick :) And when everything looks good to you then I'll take a look one more time. Could you also get someone else at AMD to look to make sure the pass option approach is fine? Maybe Matt again.

yxsamliu commented:
> Sure go ahead and cherry-pick :) And when everything looks good to you then I'll take a look one more time. Could you also get someone else at AMD to look to make sure the pass option approach is fine? Maybe Matt again.

@YonahGoldberg I cherry-picked your refactor onto this PR as `3718b15b7a2a`, with a small follow-up `77a83c08dce5` that reverts two DebugInfo test CHECKs back to upstream form. All SROA lit tests pass locally, and the new-PM defaults/print-pipeline tests are clean. Premerge CI is running now. 

@arsenm could you take a look at the pass-option approach when you have a moment? Yonah added a per-pass `canonicalize-struct-to-vector` flag on `SROAOptions` and enabled it only at the late SROA after memcpyopt, which avoids the regressions reviewers flagged earlier.

Thanks both.

YonahGoldberg committed:
[SROA] Gate struct-to-vector canonicalization on a per-pass flag

Add a `canonicalize-struct-to-vector` option to `SROAOptions`, off by default.
Only the late SROA passes in `addVectorPasses` and the two NVPTX legacy-PM
SROAs enable it, so canonicalization runs after `MemCpyOptPass`. Running it
earlier can hide memcpy chains from memcpyopt or emit wide stores with undef
suffix lanes (see the two new SROA tests). With firing gated, the helper is
simplified to require memory-intrinsic-only users, and the element-shape rule
accepts any homogeneous element count, any integer width, any FP, and integral
pointers. Opt in via `opt -passes='sroa<canonicalize-struct-to-vector>'`.

yxsamliu committed:
[SROA] Revert three tests to upstream form

The default `sroa` pass no longer canonicalizes homogeneous struct partitions
to fixed vectors, so restore the scalar CHECKs in `sroa-alloca-offset.ll`,
`user-memcpy.ll`, and `nullptr.cl` to match what the default pipeline now
produces. NVPTX still opts in, so `lower-byval-args.ll` keeps its
vector-shaped CHECKs.

arsenm commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5086,6 +5089,67 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+/// partition, SROA would select that type.
+///
+/// Applying this transformation too early may hinder memcpyopt, which may
+/// generate better code when eliminating allocas. For example, see
```
This seems like a solvable problem in memcpyopt?

YonahGoldberg replied:
I'm not really sure, if you look at `struct-to-vector-before-memcpyopt.ll`, the motivating example is:

```
memcpy tmp, obj, 16
memset obj + 16, 0, 16

; I think the above is in a separate function when the problem happens

memcpy swap.tmp, tmp, 16
memcpy tmp, other, 16
memcpy other, swap.tmp, 16
```

So the swap.tmp gets promoted in SROA and you get:

```
swap.tmp = load tmp
memcpy tmp, other, 16
store other, swap.tmp
```

The problem is that now swap.tmp is an SSA value, so memcpyopt doesn't find that swap.tmp isn't necessary because the same value is stored in `obj`.

If you allow memcpyopt to run without SROA you get:

```
memset obj + 16, 0, 16
memcpy tmp, other, 16
memcpy other, obj, 16
```

Fixing this in memcpyopt after SROA would be really hard. You'd have to do something like converting load/store pairs back to memcpys, which seems pretty brittle. Better to just leave memcpys in the IR until after memcpyopt.

yxsamliu replied:
@arsenm Adding to Yonah's reply: the general SROA <-> memcpyopt relationship is already a back-and-forth — each pass works best on its own preferred IR shape, and the pipeline runs them multiple times to let them clean up after each other. What this PR adds is a new SROA trick (promote memcpy-only struct allocas to a vector SSA value) that, when run too early, breaks the memcpy chain that memcpyopt was about to simplify. Pushing the trick to run after memcpyopt preserves the existing handoff: memcpyopt does its memcpy-level reasoning first, then late SROA picks up whatever memcpyopt couldn't eliminate. Teaching memcpyopt to reconstruct memcpys from arbitrary load/store pairs would help, but it's a much larger change with its own correctness surface, and not specific to this PR.

arsenm commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5086,6 +5089,67 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+/// generate better code when eliminating allocas. For example, see
+/// `struct-to-vector-fp-store-only-tail.ll`, which demonstrates that applying
+/// this before memcpyopt can initialize previously uninitialized memory when
+/// the alloca gets promoted to an SSA value. For another example, see
```
Is this about uninitialized memory being undef, and not poison? I don't see how changing the type fundamentally changes what bytes are defined or not

YonahGoldberg replied:
The problem is when a partition that is partially uninitialized gets promoted to SSA values. This PR allows promoting in some cases allocas that only have memcpy users. These allocas might be partially uninitialized.  If you leave it as a memcpy of partially uninitialized memory, then memcpyopt seems to recognize that and shorten the width of the memcpy and you get a `store float` If you promote, I observed that you can get something like `store <4 x float> <0, undef, undef, undef>` and you get a 16-byte store.

This does have to do with uninitialized memory being treated as undef, even though it should be poison.

https://github.com/llvm/llvm-project/pull/194042 would fix that issue, but the transformation is technically illegal. I think better to avoid initializing uninitialized memory in the first place.

arsenm commented on llvm/test/Transforms/SROA/struct-to-vector-fp-store-only-tail.ll:
Use named values in tests

yxsamliu replied:
will do

arsenm commented on llvm/test/Transforms/SROA/struct-to-vector-fp-store-only-tail.ll:
```suggestion
```

yxsamliu replied:
will do

arsenm commented on llvm/test/Transforms/SROA/struct-to-vector-before-memcpyopt.ll:
Should a variant of this be added to test/PhaseOrdering to test the pass pipeline config change?

YonahGoldberg replied:
Yeah I agree. I think we should just move this test into that folder. Didn't know that existed.

yxsamliu replied:
will do

arsenm commented on llvm/include/llvm/Transforms/Scalar/SROA.h:
```diff
@@ -21,15 +21,26 @@ namespace llvm {
+  enum CFGOption { ModifyCFG, PreserveCFG };
+
+  CFGOption CFG;
+  bool CanonicalizeStructToVector;
```
I think the name here could use work, don't really need the "Canonicalize" part

YonahGoldberg replied:
Yeah maybe just "StructToVector" is better.

yxsamliu replied:
will do

arsenm reviewed (commented):
I didn't follow the whole story here. Introducing the flag for the pass pipeline seems OK, though I feel like this is avoiding solving problems in other places. I would expect SROA to avoid introducing store of undefined values, but I would also expect the rest of the pass pipeline to deal with partially undefined stores.

(side point, I went to look at memcpyopt to see if the problem was obvious and noticed https://github.com/llvm/llvm-project/issues/197683)

YonahGoldberg reviewed (commented):
@yxsamliu let me know what you think.

efriedma-quic commented:
How does this interact with https://discourse.llvm.org/t/rfc-plan-for-eliminating-allocainst-getallocatedtype-leaving-only-getarraysize/89969 ?

YonahGoldberg commented:
> How does this interact with https://discourse.llvm.org/t/rfc-plan-for-eliminating-allocainst-getallocatedtype-leaving-only-getarraysize/89969 ?

I believe this was already addressed above (if you scroll really far up) but I'll add my thoughts. This optimization addresses struct allocas where the only users are memory intrinsics (memcpy, memset, etc...). Currently SROA doesn't touch these because the partition type will default to the struct (the type of the alloca). But if you choose a vector partition type then we can promote the alloca to an SSA value.

But the optimization is more general and doesn't need to depend on the type of the alloca. First of all, this is only necessary for struct allocas and (in a future follow-up PR) array allocas. Scalar allocas will default to a scalar partition type when they only have memory intrinsic users and these will get promoted. I plan to put up a follow-up PR for array allocas after this goes in, which should fix this Julia compiler issue: https://github.com/llvm/llvm-project/issues/164308.

But in a future world where there are only i8 array allocas (i.e. the struct allocas get canonicalized to i8 array allocas by instcombine), then you can just rely on the code that promotes these i8 array alloca to i8 vector SSA values.

In fact, we could stop looking at the type of the alloca right now. We could just say that if there is an alloca with only memory intrinsic users and the size of the alloca is N bytes, then choose `<N x i8>` for the partition type. Then the alloca will get promoted to an <N x i8> SSA value. Is the new byte type maybe even a better choice? The only thing is I would worry that for something like `struct {i32, i32, i32, i32}` if you promote this as `<16 x i8>` instead of `<4 x i32>`, maybe the CodeGen is worse for large byte vectors.

What do you think about the CodeGen issue from your perspective working on Aarch64?

Maybe @vtjnash has thoughts.

efriedma-quic commented:
I think if we're going to start ignoring the type information at some point, we should do it sooner rather than later, to avoid churn.

It probably makes sense on most targets for SROA to turn memory into a register even if there aren't any non-memcpy accesses on CPU targets.  I'm not sure if we'd want to use vector types on AArch64; integer registers usually have lower latency.  But I'm not sure how much the exact type SROA uses matters; hopefully by the time we get to codegen, the backend can fix up the type.  (Well, except that we need to use a byte type for correctness if the memory contains any pointers.)

YonahGoldberg commented:
> I'm not sure if we'd want to use vector types on AArch64; integer registers usually have lower latency.

Well for large memcpys if you promote I think you have to promote to a vector type.

> But I'm not sure how much the exact type SROA uses matters; hopefully by the time we get to codegen, the backend can fix up the type.

Maybe, we could try benchmarking on a couple backends to see if making the type i8 always is fine. @yxsamliu thoughts?

> (Well, except that we need to use a byte type for correctness if the memory contains any pointers.)

Is this true? Like in general is it not legal for me to promote:

```
memcpy B, A
memcpy C, B
```

to:

```
B = load A
C = store B
```

If the alloca for B is `alloca struct {ptr, ptr}` and for the SSA value of B I choose `<8 x i8>`? If so, should we choose `<8 x byte>` for the SSA value? I haven't followed along with how well-handled the byte type is.

Note: I know this is better optimized as `memcpy C, A`, I'm just using this as a simple example.

efriedma-quic commented:
>  If so, should we choose `<8 x byte>` for the SSA value?

Well, what matters is the contents of the memory, not the declared type... but in theory we need `<8 x b8>`, yes.  Byte types are still a work in progress, though; the implementation is still missing pieces.

efriedma-quic commented:
> Well for large memcpys if you promote I think you have to promote to a vector type.

We can use iN with large N.  It gets a little sketchy if N is more than 512, but vectors also have limitations at sufficiently large sizes.

vtjnash commented:
You can't use iN if any of the stored bits might have been poison, since that propagates to the whole value even if they don't get used. i8 vectors or byte types avoid that issue

yxsamliu commented:
> > I'm not sure if we'd want to use vector types on AArch64; integer registers usually have lower latency.
> 
> Well for large memcpys if you promote I think you have to promote to a vector type.
> 
> > But I'm not sure how much the exact type SROA uses matters; hopefully by the time we get to codegen, the backend can fix up the type.
> 
> Maybe, we could try benchmarking on a couple backends to see if making the type i8 always is fine. @yxsamliu thoughts?
> 
@efriedma-quic On the RFC angle and Yonah's "always i8" question: I gave it a quick try on the AMDGPU kernel that originally motivated this PR and on the 50-sample llvm-opt-benchmark set we've been using as the most sensitive regression coverage for this change, and didn't see any observable downstream difference between `<N x EltTy>` and `<N x i8>` — same SROA decisions, same post-O3 IR, same asm. So the RFC-aligned direction of dropping the alloca type and just picking a byte vector by size looks viable here, and a natural follow-up that could subsume
the current homogeneous-struct heuristic and extend it to heterogeneous allocas. For this PR I'd lean toward keeping `<N x EltTy>` as the conservative default for now, since it preserves the natural type when one is available, and switch over once the byte-type story (per vtjnash's poison note) is in better shape.

yxsamliu committed:
[SROA] Rename CanonicalizeStructToVector option to StructToVector

Per @arsenm review feedback on PR #165159: drop the redundant
"Canonicalize" verb from the pass-option name. Renames the SROAOptions
field, the pipeline-parser string ("canonicalize-struct-to-vector" ->
"struct-to-vector"), the createSROAPass parameter, all call sites, and
the existing RUN-line strings in two SROA tests.

yxsamliu committed:
[SROA] Clean up struct-to-vector-fp-store-only-tail.ll

Per @arsenm review feedback on PR #165159:
 - Rename numeric SSA values (%1, %2, %3) to named values (%dst, %src,
   %src.tail) so the test reads more clearly and won't churn on future
   edits. Regenerated CHECK lines via update_test_checks.py.
 - Drop the auto-emitted "Function Attrs:" comment line above the
   llvm.memcpy declaration; the attrs already appear on the declare
   itself.

yxsamliu committed:
[SROA] Move struct-to-vector-before-memcpyopt.ll to test/PhaseOrdering

Per @arsenm review feedback on PR #165159: the test exists to verify
that struct-to-vector canonicalization runs only after memcpyopt in the
default pipeline, which is a pipeline-configuration concern rather than
a SROA-pass unit test. Move it under test/PhaseOrdering/ and switch the
single RUN line to "opt -passes='default<O3>'" so it exercises the real
pipeline directly instead of a synthetic pass list.

yxsamliu committed:
[SROA] Rename StructToVector option to AggregateToVector

Per @YonahGoldberg follow-up on PR #165159: the transformation is
expected to extend to array allocas in a follow-up (motivated by
issue #164308 for the Julia frontend), so the per-pass option should
not be tied to "struct". Renames the SROAOptions field, the
pipeline-parser string ("struct-to-vector" -> "aggregate-to-vector"),
the createSROAPass parameter, all call sites, and the existing RUN-line
strings in the three SROA tests.

yxsamliu committed:
[SROA] Rename StructToVector option to AggregateToVector

Per @YonahGoldberg follow-up on PR #165159: the transformation is
expected to extend to array allocas in a follow-up (motivated by
issue #164308 for the Julia frontend), so the per-pass option should
not be tied to "struct". Renames the SROAOptions field, the
pipeline-parser string ("struct-to-vector" -> "aggregate-to-vector"),
the createSROAPass parameter, all call sites, and the existing RUN-line
strings in the three SROA tests.

YonahGoldberg reviewed (approved):
LGTM

YonahGoldberg commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5086,6 +5088,67 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
+/// the alloca gets promoted to an SSA value. For another example, see
+/// `struct-to-vector-before-memcpyopt.ll`, which demonstrates that applying
+/// this before memcpyopt can result in promoting an alloca so that we load a
+/// tempory value instead of copying the tempory value into memory, whereas
```
Nit: s/tempory/temporary

yxsamliu replied:
done

bjope commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5086,6 +5088,67 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
 
+/// Try to canonicalize a homogeneous struct partition to a vector type.
+///
+/// We can do this if all the elements of the struct are the same and tightly
```
Consider IR like this:

```
define void @d2(ptr %c) {
entry:
  %e = alloca { i5, i5 }, align 1
  call void @llvm.memcpy.p0.p0.i32(ptr align 1 %e, ptr align 1 %c, i32 2, i1 true)
  ret void
}

define void @d3(ptr %c) {
entry:
  %e = alloca { i5, i5, i5 }, align 1
  call void @llvm.memcpy.p0.p0.i32(ptr align 1 %e, ptr align 1 %c, i32 3, i1 true)
  ret void
}
```
The first function wil be changed to use <2 x i5>, but the second if not changed to use <3 x i5>.

Not sure if this is a problem really, but the code comment here talks about restricting to cases when the elements of the struct are being tightly packed. But that is not really what is happening when comparing the size of the struct and the size of the vector.

Maybe there should be a check that DataLayout::typeSizeEqualsStoreSize is true for the element type if we want to restrict this to structs without padding?

YonahGoldberg replied:
Thank you! I think this is a real bug. Please see my fix at: https://github.com/llvm/llvm-project/pull/201967.
If you have a chance, do you also want to look at this: https://github.com/llvm/llvm-project/pull/201434, which fixes another issue?

antoniofrighetto commented on llvm/test/Transforms/SROA/struct-to-vector.ll:
```diff
@@ -0,0 +1,389 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
+; RUN: opt -passes='sroa<aggregate-to-vector>,gvn,instcombine,simplifycfg' -S %s | FileCheck %s
```
Why are we also testing `gvn,instcombine,simplifycfg` here? Is this needed?

YonahGoldberg replied:
Forgot about this, TODO to update
